File: web/src/api/update.ts
import client from './client';

/**
 * 版本检查响应 —— 字段逐字对齐后端 internal/api/update.go 的 Check handler。
 * 与 /api/v1/system/* 一致采用 snake_case。
 *
 * 必返字段：current / frp / deployment_mode / self_update_enabled /
 *           has_update / can_self_update / reason
 * 仅成功时返回：latest / changelog / html_url / published_at
 * 仅失败时返回：check_error
 */
export interface VersionCheckResp {
  current: string;
  frp?: string;
  deployment_mode: string; // docker | systemd | openrc | launchd | windows-service | manual
  self_update_enabled: boolean;
  has_update: boolean;
  can_self_update: boolean;
  reason: string;
  latest?: string;
  changelog?: string;
  html_url?: string;
  published_at?: string;
  check_error?: string;
}

/** POST /api/v1/system/update 的 202 响应体。 */
export interface UpdateStartResp {
  status: string;
  from: string;
  to: string;
  message: string;
}

/** 检查最新版本（force=true 绕过后端 ~1h 缓存）。 */
export async function checkVersion(force = false): Promise<VersionCheckResp> {
  const resp = await client.get<VersionCheckResp>('/api/v1/version/check', {
    params: force ? { force: 1 } : undefined,
  });
  return resp.data;
}

/** 触发一键更新（force=true 即使已是最新也强制重装）。返回后服务即将重启。 */
export async function startUpdate(force = false): Promise<UpdateStartResp> {
  const resp = await client.post<UpdateStartResp>('/api/v1/system/update', null, {
    params: force ? { force: 1 } : undefined,
  });
  return resp.data;
}

export const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

/** 读取当前自更新日志 (update.log)，供更新中实时展示步骤。重启窗口期会失败，调用方需容忍。 */
export async function fetchUpdateLog(): Promise<string> {
  const resp = await client.get<{ content?: string }>('/api/v1/system/update/log', { timeout: 4000 });
  return resp.data?.content || '';
}

/**
 * 轮询 /api/v1/version，直到 daemon 版本不再等于 fromVersion（即新进程已起来），
 * 或超时。重启期间连接会失败，这里容忍并继续轮询。
 * 返回新的 daemon 版本字符串；超时抛错。
 */
export async function waitForRestart(fromVersion: string, timeoutMs = 120000): Promise<string> {
  const start = Date.now();
  // 给旧进程一点时间开始关闭，避免一上来就读到旧版本
  await sleep(3000);
  while (Date.now() - start < timeoutMs) {
    try {
      const resp = await client.get('/api/v1/version', { timeout: 4000 });
      const v: string = resp.data?.daemon || resp.data?.version || '';
      if (v && v !== fromVersion) return v;
    } catch {
      // 重启期间的连接失败属预期，继续轮询
    }
    await sleep(2500);
  }
  throw new Error('等待服务重启超时');
}

File: internal/api/update.go
package api

import (
	"log/slog"
	"net/http"

	"github.com/nue-mic/kwrt-net-manager/internal/selfupdate"
	"github.com/nue-mic/kwrt-net-manager/pkg/version"
)

// UpdateHandler serves the version-check and self-update endpoints:
//
//	GET  /api/v1/version/check  — compare running vs latest GitHub release
//	POST /api/v1/system/update  — launch a detached in-place upgrade + restart
type UpdateHandler struct {
	updater      *selfupdate.Updater
	selfUpdateFn func() bool
	logger       *slog.Logger
}

// NewUpdateHandler wires an UpdateHandler. selfUpdateFn reports whether the
// web-triggered self-update is currently enabled (env default, overridable at
// runtime via the system-config UI); when false the POST endpoint is refused.
func NewUpdateHandler(dataDir string, selfUpdateFn func() bool, logger *slog.Logger) *UpdateHandler {
	return &UpdateHandler{
		updater: selfupdate.New(selfupdate.Config{
			DataDir: dataDir,
		}),
		selfUpdateFn: selfUpdateFn,
		logger:       logger,
	}
}

// Check reports the current version alongside the latest GitHub release, the
// detected deployment mode and whether a one-click self-update is possible.
// Pass ?force=1 to bypass the ~1h cache.
func (h *UpdateHandler) Check(w http.ResponseWriter, r *http.Request) {
	force := r.URL.Query().Get("force") == "1" || r.URL.Query().Get("refresh") == "1"

	mode := selfupdate.DetectDeployment()
	canDeploy, reason := selfupdate.CanSelfUpdate(mode)

	out := map[string]any{
		"current":             version.Number,
		"deployment_mode":     string(mode),
		"self_update_enabled": h.selfUpdateFn(),
	}

	rel, err := h.updater.CheckLatest(r.Context(), force)
	if err != nil {
		out["has_update"] = false
		out["can_self_update"] = false
		out["check_error"] = err.Error()
		out["reason"] = "无法获取最新版本：" + err.Error()
		WriteJSON(w, http.StatusOK, out)
		return
	}

	out["latest"] = rel.Tag
	out["changelog"] = rel.Changelog
	out["html_url"] = rel.HTMLURL
	out["published_at"] = rel.PublishedAt
	out["has_update"] = selfupdate.HasUpdate(version.Number, rel.Tag)

	// can_self_update is the capability (deployment supports it AND operator
	// enabled it); the frontend combines it with has_update to enable the
	// button. reason explains a disabled state.
	canSelf := canDeploy && h.selfUpdateFn()
	if !h.selfUpdateFn() {
		reason = "管理员已禁用 Web 端自更新（KWRTNET_SELF_UPDATE_ENABLED=false）"
	}
	out["can_self_update"] = canSelf
	out["reason"] = reason

	WriteJSON(w, http.StatusOK, out)
}

// Update launches the detached self-update and returns 202 immediately. The
// daemon is about to be restarted by the updater, so the client should poll
// /health and /version until the version changes. Pass ?force=1 to reinstall
// the current latest even when already up to date.
func (h *UpdateHandler) Update(w http.ResponseWriter, r *http.Request) {
	if !h.selfUpdateFn() {
		WriteError(w, http.StatusForbidden, CodeForbidden,
			"Web 端自更新已禁用（KWRTNET_SELF_UPDATE_ENABLED=false）", nil)
		return
	}

	mode := selfupdate.DetectDeployment()
	if ok, reason := selfupdate.CanSelfUpdate(mode); !ok {
		WriteError(w, http.StatusBadRequest, CodeInvalidState, reason, nil)
		return
	}

	// Resolve the target so the updater installs a deterministic version and
	// we can echo it back.
	rel, err := h.updater.CheckLatest(r.Context(), false)
	if err != nil {
		WriteError(w, http.StatusBadGateway, CodeUpstreamFailure,
			"无法获取最新版本："+err.Error(), nil)
		return
	}

	force := r.URL.Query().Get("force") == "1"
	if !selfupdate.HasUpdate(version.Number, rel.Tag) && !force {
		WriteError(w, http.StatusConflict, CodeConflict,
			"已是最新版本 "+version.Number+"（如需强制重装请加 ?force=1）", nil)
		return
	}

	// Start the log fresh so the web UI streams only this run's steps.
	h.updater.ResetLog(version.Number, rel.Tag)

	if err := h.updater.StartUpdate(rel.Tag); err != nil {
		h.logger.Error("self-update spawn failed", "err", err)
		WriteError(w, http.StatusInternalServerError, CodeInternal,
			"启动更新失败："+err.Error(), nil)
		return
	}

	h.logger.Warn("self-update initiated; service will restart",
		"from", version.Number, "to", rel.Tag, "mode", string(mode))
	WriteJSON(w, http.StatusAccepted, map[string]any{
		"status":  "updating",
		"from":    version.Number,
		"to":      rel.Tag,
		"message": "更新已开始，服务即将重启，请稍候…",
	})
}

// Log returns the current self-update log (update.log) so the web UI can show
// the live progress steps while an update runs.
func (h *UpdateHandler) Log(w http.ResponseWriter, r *http.Request) {
	WriteJSON(w, http.StatusOK, map[string]any{"content": h.updater.ReadLog()})
}
